packages/server/src/arpa_reporter/db/arpa-subrecipients.js

```diff
@@ -72,21 +60,54 @@ async function getRecipient(id, trns = knex) {
         .then((rows) => rows[0]);
 }
 
-async function findRecipient(uei = null, tin = null, trns = knex) {
+async function findRecipient(fieldType = null, value = null, trns = knex) {
     const tenantId = useTenantId();
     const query = baseQuery(trns).where('arpa_subrecipients.tenant_id', tenantId);
 
-    if (uei) {
-        query.where('arpa_subrecipients.uei', uei);
-    } else if (tin) {
-        query.where('arpa_subrecipients.tin', tin);
+    if (fieldType === 'uei') {
+        query.where('arpa_subrecipients.uei', value);
+    } else if (fieldType === 'tin') {
+        query.where('arpa_subrecipients.tin', value);
+    } else if (fieldType === 'name') {
+        query.where('arpa_subrecipients.name', value);
     } else {
         return null;
     }
 
     return query.then((rows) => rows[0]);
 }
 
+async function createRecipient(recipient, trns = knex) {
+    const tenantId = useTenantId();
+    if (!(recipient.uei || recipient.tin || recipient.name)) {
+        throw new Error('recipient row must include a `uei`, `tin`, or `name` field');
+    }
+
+    if (recipient.uei) {
+        const existingRecipient = await findRecipient('uei', recipient.uei, trns);
+        if (existingRecipient) {
+            throw new Error('A recipient with this UEI already exists');
+        }
+    }
+    if (recipient.tin) {
+        const existingRecipient = await findRecipient('tin', recipient.tin, trns);
+        if (existingRecipient) {
+            throw new Error('A recipient with this TIN already exists');
+        }
+    }
+    if (!recipient.tin && !recipient.uei && recipient.name) {
+        const existingRecipient = await findRecipient('name', recipient.name, trns);
```

Comment:
Can the same name exist across multiple tenants?

Reply:
yes, `findRecipient` scopes by tenant by default, so we should only be looking for duplicates within a tenant.